### PR TITLE
Texture cache fixes

### PR DESCRIPTION
**Pull request by ruipin**
While investigating the P5 texture swap issues, I stumbled into two different bugs in the texture cache, specifically in the `overlaps_page` and `get_intersecting_set` methods.

This pull request fixes these two bugs:

- In `overlaps_page`, if `address` is not page-aligned, the calculated end address for the corresponding page is incorrect which can lead to false positives, causing some textures to be re-uploaded unnecessarily.
- When the trampled range changes, `get_intersecting_set` restarts the outer loop. However, due to an off-by-one error, it skips the first cache entry when doing so. This can cause a texture to not be correctly unlocked, which could lead to issues such as texture corruption or even deadlocks.

I have ran these through @kd-11 and he has confirmed these are actual bugs.

In addition, after some discussion with @kd-11, I have refactored `get_intersecting_set` to avoid unnecessary looping. It will now only execute the minimum number of iterations necessary to ensure all textures have been processed using the correct `trampled_range`. I have tested this last change by comparing it directly with the old implementation (see https://github.com/ruipin/rpcs3/commit/25b6f2f172ef8d187ae9d0886367690a4c8bfc19) and saw no difference in behaviour.

Note that I have only tested these changes with P5, since that is the only game I have available right now.

(This is my first PR here, comments/suggestions/criticism/etc are welcome!)

### CODE REVIEW

**Comment by ruipin** [comment]
I've restored the ordering of the variable declarions in `get_intersecting_set` at @kd-11's request.

**Comment by ruipin** [comment]
I have added a third fix to this PR. After a lot more time reviewing the texture cache code, and after pestering @kd-11 a lot with many questions (thank you for your help!) I found another bug:

- `invalidate_range_impl_base` does not mark all textures that will only be
unprotected as dirty when doing a deferred flush, since that is done by
`flush_all`. However, if there are no sections to flush, the deferred flush will
use the same code path as non-deferred flushes for unprotecting textures
and forget to mark them as dirty.

With this fix I believe I have finally tracked down the bug causing the P5 texture swaps, but further testing is needed.

**Comment by ruipin** [comment]
I have updated #4298 with some test results for this PR.

tl;dr: P5 texture swap bug is fixed during normal gameplay, but not inside the Velvet Room. I suspect there is a fourth bug somewhere that only triggers in there.
